include/swift/AST/DiagnosticsSema.def

```diff
@@ -64,6 +64,9 @@ ERROR(ambiguous_reference_to_decl,none,
 ERROR(ambiguous_subscript,none,
       "ambiguous subscript with base type %0 and index type %1",
       (Type, Type))
+NOTE(ambiguous_subscript_candidate_requires_operator,none,
+     "candidate requires use of the subscript operator", ())
```

Comment:
@jrose-apple I'm not really happy with this message, could you please help me to phrase this better? The problem is that once name "subscript" is turned into subscript operator it could be ambiguous like in one of the examples (e.g. based on result type), but I don't only want to point out ambiguity but also mention that fixing it would require using subscript operator as well...

Reply:
I feel like the error message itself is misleading, and the note doesn't cover enough. But in cases where you're going to have a note, maybe this can replace "found this candidate" instead of being emitted in addition to it?

(The reason this matters at all is because ```func `subscript`(_ sender: Any?)``` exists.)

Reply:
Unfortunately we aren't there yet to replace default "ambiguous member" diagnostic for multiple solutions with something better and add fix-its, so I can only provide additional information with notes and specially I only use this note if the candidate we found has to use subscript operator fix e.g.

```swift
struct S {
  subscript(_ x: String) -> Int { return 42 }
  subscript(_ x: String) -> String { return "42" }
}

func foo(_ s: S) {
  _ = s.subscript("ultimate question")
}
```

Is going to produce error like this now:

```
error: ambiguous reference to subscript 'subscript(_:)'
    _ = s.subscript("ultimate question")
          ^
note: candidate requires use of the subscript operator
    subscript(_ x: String) -> Int { return 42 }
    ^
note: candidate requires use of the subscript operator
    subscript(_ x: String) -> String { return "42" }
    ^
```

> (The reason this matters at all is because func `subscript`(_ sender: Any?) exists.)

Sorry I don't understand what you mean completely... When such overload exists fix that I added would not be attempted.

Reply:
> Sorry I don't understand what you mean completely... When such overload exists fix that I added would not be attempted.

We probably already have tests for this, so I won't worry about it.

I'll conclude that I don't think it's worth doing this work until we can change the other diagnostics, then. I find the new top-level errors *less* clear than the old ones.

Reply:
The reason why I'm doing this now is because I'm trying to get missing members to be diagnosed via fixes, this is one of the things which came up just like `member used as a property` error. I think that reporting ambiguity here is an improvement as well as couple of other errors related to incorrect arguments where it used to just diagnose as `use subscript operator`, because once I make it so re-labeling is diagnosed via fixes we'd get a nice multi-error suggestion with fix-its instead of current chain of errors.

Reply:
How do you determine which is the main problem? Because if I replace `subscript` with subscript operator it would result in ambiguous error, or if I fix ambiguity by adding contextual type it would then result in error about subscript misuse...

It would  probably be best to attach fix-it to each note, but I don't believe we can do that at the moment. At least new ambiguous error combines information about both problems - ambiguity + need to use subscript operator.

Reply:
I don't think this change contradicts what you are saying, original behavior is going to be preserved - your example is still going to diagnose as `cannot convert value of type 'String' to expected argument type 'Int'`. The only way to produce ambiguous diagnostic - if there were no overloads for `subscript` name and overloads of the subscript operator produce solutions which we can't disambiguate.

Reply:
Right. So, given the existing behavior when there *are* methods named 'subscript', the spelling of the call should be considered more important than the types of the arguments matching up.

Reply:
Wait, I'm confused - in previous comment you mentioned that in you'd expect example to produce argument related error (which is what it currently does), but now you are saying that it should actually produce suggestion to replace name with operator?

Reply:
I'm sorry if I've been unclear. I've been trying to state that "you wrote `subscript` when you should have used brackets" is the error you should be told about first.

Reply:
I understood that, but it contradicts your previous comment given its example, also current behavior is consistent with it and is unchanged by this fix.

Reply:
Ok, I think I understand what you mean now so the actual diagnostic then would be something like: 

```swift
error: value of type 'SubscriptTest1' has no property or method named 'subscript'; did you mean to use the subscript operator?

note: found this candidate:
    subscript(_ x: String) -> Int { return 42 }
    ^
note: found this candidate:
    subscript(_ x: String) -> String { return "42" }
    ^
```

Reply:
Done!